lib/rucio/transfertool/fts3.py

```diff
@@ -150,52 +151,45 @@ def submit(self, files, job_params, timeout=None):
             expected_transfer_id = self.__get_deterministic_id(job_params["sid"])
             logging.debug("Submit bulk transfers in deterministic mode, sid %s, expected transfer id: %s", job_params["sid"], expected_transfer_id)
 
-        # bulk submission
-        params_dict = {'files': files, 'params': job_params}
-        params_str = json.dumps(params_dict, cls=APIEncoder)
+        # Multi_VO mode
+        if self.multi_vo:
+            vo_list = []
+            vo_sorted_files = {}
 
-        post_result = None
-        try:
-            start_time = time.time()
-            post_result = requests.post('%s/jobs' % self.external_host,
-                                        verify=self.verify,
-                                        cert=self.cert,
-                                        data=params_str,
-                                        headers=self.headers,
-                                        timeout=timeout)
-            record_timer('transfertool.fts3.submit_transfer.%s' % self.__extract_host(self.external_host), (time.time() - start_time) * 1000 / len(files))
-            labels = {'host': self.__extract_host(self.external_host)}
-            SUBMISSION_TIMER.labels(**labels).observe((time.time() - start_time) * 1000 / len(files))
-        except ReadTimeout as error:
-            raise TransferToolTimeout(error)
-        except JSONDecodeError as error:
-            raise TransferToolWrongAnswer(error)
-        except Exception as error:
-            logging.warning('Could not submit transfer to %s - %s' % (self.external_host, str(error)))
+            for transfer_file in files:
+                vo_name = job_params['account'].vo
+                if vo_name not in vo_list:
+                    vo_list.append(vo_name)
+                    vo_sorted_files[vo_name] = []
+                vo_sorted_files[vo_name].append(transfer_file)
+
+            for vo in vo_list:
+
+                sorted_files = vo_sorted_files[vo]
+
+                # bulk submission
+                params_dict = {'files': sorted_files, 'params': job_params}
+                params_str = json.dumps(params_dict, cls=APIEncoder)
+
+                post_result = None
+
+                vo_cert = config_get('vo_certs', vo, False, config_get('conveyor', 'usercert', False, None))
+
+                self.cert = (vo_cert, vo_cert)
+
+                transfer_id = self.__attempt_transfer(params_str, timeout, files, expected_transfer_id, post_result)
+                return transfer_id
```

Comment:
This will stop submission after handling the first VO.

Also, now with the request code out of the way moved to a separate function, it feels like the vo handling code must be done upstream. Somewhere in bulk_group_transfers or in bulk_submit transfers. The 
```
self.cert = (vo_cert, vo_cert) 
```
part look like looking for troubles. Somebody re-using the same transfertool instance may not expect for certificates to have some random value.

Reply:
So it would stop the submission, good catch! I will fix that now

As for the
`self.cert = (vo_cert, vo_cert) `
I was confused about that as well, but that seemed to be how it was handeled  for non-multiVO instances

        ```
if self.external_host.startswith('https://'):
            if self.token:
                self.cert = None
                self.verify = False
                self.headers['Authorization'] = 'Bearer ' + self.token
            else:
                self.cert = (usercert, usercert)
                self.verify = False
```
So was attempting to not cause issues with multiVO, or do you mean that is now called vo_cert? I suppose it can be renamed usercert like above

Reply:
I mean that the object state is changed by a function which doesn't look (without reading it code) that it will change the object state. This may be handled correctly in the current state of code, but, in the future, somebody can decide to add an fts.query() after an fts.submit(). That code will pass the review because nobody would even think to look what submit() does inside. Will be deployed. And will fail on release (but only for people using multi-vo). That's why I say that it "feels" like looking for trouble, not that it is wrong right now.

Reply:
Also I have done some more testing with my modified test containers (I altered the multi_vo_tests_docker.sh to initialise both VOs and and add files, containers, and datasets for each, and my testing does the submission for both VOs

I also have the look up there as its where the certificate is looked up currently is. so made the most sense. But I am happy to move it else where. but then would need to pass the cert down to the FTS

Reply:
It may have worked because you did run the submitter with the default value for group_bulk = 1. If you increase this parameter, I don't thin it will work. Because of that return

Reply:
Also, would be nice to add an automatic test. You can start from one of the existing ones in test_conveyor_submitter.py . You can also check test_reaper.py for some examples of multi-vo testing

Reply:
I have been thinking about your comment about VO handelling should be done in bulk_group_transfers
Which I think is a really good idea and stream lines a lot of the code if I can (as the current bulk process sorts by external_host) sort by VO prior to the sort by external_host, this will allow for a small change in the parsing of the grouped_jobs later in the submit, and submit all the transfers per external_host with the correct usercert

This of course is a different approach from what I have currently and I am coding and will be testing that version seperately but may be a better solution.

I also agree that I should look into writing some automatic testing for this portion of the code for multi-vo

Thanks very much for the comments and feedback